apps/opencs/view/render/terrainshapemode.cpp

```diff
@@ -0,0 +1,1459 @@
+        {
+            limitAlteredHeights(cellCoordinates);
+        }
+        std::reverse(mAlteredCells.begin(), mAlteredCells.end()); //Instead of alphabetical order, this should be fixed to sort cells by cell coordinates
+        for(CSMWorld::CellCoordinates cellCoordinates: mAlteredCells)
+        {
+            if (!limitAlteredHeights(cellCoordinates, true)) passing = false;
+        }
+        ++passes;
+        if (passes > 2)
+        {
+            Log(Debug::Warning) << "Warning: User edit exceeds accepted slope steepness. Automatic limiting has failed, edit has been discarded.";
+            if (CSVRender::PagedWorldspaceWidget *paged =
+                dynamic_cast<CSVRender::PagedWorldspaceWidget *> (&getWorldspaceWidget()))
+            {
+                paged->resetAllAlteredHeights();
+                mAlteredCells.clear();
+                return;
+            }
+        }
+    }
+}
+
+void CSVRender::TerrainShapeMode::applyTerrainEditChanges()
+{
+    CSMDoc::Document& document = getWorldspaceWidget().getDocument();
+    CSMWorld::IdTable& landTable = dynamic_cast<CSMWorld::IdTable&> (
+        *document.getData().getTableModel (CSMWorld::UniversalId::Type_Land));
+    CSMWorld::IdTable& ltexTable = dynamic_cast<CSMWorld::IdTable&> (
+        *document.getData().getTableModel (CSMWorld::UniversalId::Type_LandTextures));
+
+    int landshapeColumn = landTable.findColumnIndex(CSMWorld::Columns::ColumnId_LandHeightsIndex);
+    int landMapLodColumn = landTable.findColumnIndex(CSMWorld::Columns::ColumnId_LandMapLodIndex);
+    int landnormalsColumn = landTable.findColumnIndex(CSMWorld::Columns::ColumnId_LandNormalsIndex);
+
+    QUndoStack& undoStack = document.getUndoStack();
+
+    sortAndLimitAlteredCells();
+
+    undoStack.beginMacro ("Edit shape and normal records");
+
+    for(CSMWorld::CellCoordinates cellCoordinates: mAlteredCells)
+    {
+        std::string cellId = CSMWorld::CellCoordinates::generateId(cellCoordinates.getX(), cellCoordinates.getY());
+        undoStack.push (new CSMWorld::TouchLandCommand(landTable, ltexTable, cellId));
+        const CSMWorld::LandHeightsColumn::DataType landShapePointer = landTable.data(landTable.getModelIndex(cellId, landshapeColumn)).value<CSMWorld::LandHeightsColumn::DataType>();
+        const CSMWorld::LandMapLodColumn::DataType landMapLodPointer = landTable.data(landTable.getModelIndex(cellId, landMapLodColumn)).value<CSMWorld::LandMapLodColumn::DataType>();
+        CSMWorld::LandHeightsColumn::DataType landShapeNew(landShapePointer);
+        CSMWorld::LandMapLodColumn::DataType mapLodShapeNew(landMapLodPointer);
+
+        // Generate land height record
+        for(int i = 0; i < ESM::Land::LAND_SIZE; ++i)
+        {
+            for(int j = 0; j < ESM::Land::LAND_SIZE; ++j)
+            {
+                if (CSVRender::PagedWorldspaceWidget *paged =
+                    dynamic_cast<CSVRender::PagedWorldspaceWidget *> (&getWorldspaceWidget()))
```

Comment:
I doubt optimizer can put this condition outside of loop. In this case `ESM::Land::LAND_SIZE * ESM::Land::LAND_SIZE = 4225` dynamic casts are happening here. Probably it doesn't affect performance too much and does it matters here but anyway it will be definitely slower than just one dynamic cast before the loop.